Components/Queues/AsyncQueueConsumer.php

```diff
@@ -4,90 +4,146 @@
-            $this->manager->disconnect();
-        }
-
-        return true;
     }
 
     /**
+     * Returns the consumer name.
+     *
      * {@inheritdoc}
      */
     public function getName()
     {
-        return 'async_queue_consumer';
+        return sprintf(self::CONSUMER_TAG, gethostname(), getmypid());
     }
 
+    /**
+     * Returns the queue name properly treated with queue prefix.
+     */
     protected function getQueueName(EndpointInterface $endpoint): string
     {
         $options = $endpoint->getOptions();
 
         return "{$options[QueueProtocol::OPTION_PREFIX]}{$options[QueueProtocol::OPTION_QUEUE_NAME]}";
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function cleanUp(EndpointInterface $endpoint)
+    {
+        $this->driver->disconnect();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function confirmMessage(EndpointInterface $endpoint, MessageInterface $message)
+    {
+        $this->driver->ack($message->getMessageId());
+    }
+
+    public function asyncConsume(EndpointInterface $endpoint, callable $callback)
+    {
+        $this->driver->declareChannel();
+        $queueName = $this->getQueueName($endpoint);
+
+        $this->driver->declareQueue($queueName);
+        $this->driver->consume($this->getName(), $queueName, $callback);
+    }
+
+    public function waitNoBlock()
+    {
+        if ($this->driver->isConsuming()) {
+            $this->driver->waitNoBlock();
+        }
+    }
+
+    public function wait()
+    {
+        if ($this->driver->isConsuming()) {
+            $this->driver->wait();
+        }
+    }
+
+    protected function process(EndpointInterface $queueEndpoint, QueueMessageInterface $message)
+    {
+        $endpoint = $this->smartesbHelper->getEndpointFactory()->createEndpoint($message->getDestinationURI(), EndpointFactory::MODE_CONSUME);
+        $queueEndpoint->getHandler()->handle($message->getBody(), $endpoint);
+    }
+
+    /**
+     * Overrides the main callback function to convert the AMQPMessage from the queue into a QueueMessage.
+     *
+     * {@inheritDoc}
+     */
+    public function callback(EndpointInterface $endpoint)
+    {
+        return function (AMQPMessage $message) use ($endpoint) {
+            try {
+                $start = microtime(true);
+                $queueMessage = $this->serializer->deserialize($message->getBody(), SerializableInterface::class, $this->driver->getFormat());
+                $this->consumptionDuration = (microtime(true) - $start) * 1000;
+
+                $queueMessage->setMessageId($message->getDeliveryTag());
+            } catch (\Exception $exception) {
+                // TODO Verify "headers" are passed correctly. might need to access "data" key after get_properties
+                $this->consumptionDuration = (microtime(true) - $start) * 1000;
+                $this->getExceptionHandler()($exception, ['headers' => $message->get_properties(), 'body' => $message->getBody()]);
```

Comment:
getExceptionHandler() method doesn't receive any parameter on the definition. This shouldn't be there
```
($exception, ['headers' => $message->get_properties(), 'body' => $message->getBody()])
```

Reply:
We are getting the exception handler and then `invoking` it. Refer to the ExceptionHandlerInterface for this.

Reply:
Ok, understood